virtool/uploads/utils.py

```diff
@@ -1,45 +1,81 @@
+import os
+import pathlib
+from typing import Optional, Dict
+
 import aiofiles
+import aiohttp.web
 from cerberus import Validator
-import os
+
+import virtool.utils
 
 CHUNK_SIZE = 4096
 
 
+def is_gzip_compressed(chunk: bytes) -> bool:
+    """
+    Check if a file is gzip compressed by checking the first two bytes for the gzip magic number.
+
+    :param chunk: First byte chunk from a file being uploaded
+    :return: A boolean indicating whether the file is gzip compressed or not
+    """
+    return chunk[:2] == b"\x1f\x8b"
+
+
 def naive_validator(req) -> Validator.errors:
     """
     Validate `name` given in a HTTP request using cerberus
 
     """
-    v = Validator({
-        "name": {"type": "string", "required": True}
-    }, allow_unknown=True)
+    v = Validator({"name": {"type": "string", "required": True}}, allow_unknown=True)
 
     if not v.validate(dict(req.query)):
         return v.errors
 
 
-async def naive_writer(req, file_path) -> int:
+async def naive_writer(
+    req: aiohttp.web.Request, path: pathlib.Path, compressed: bool = False
+) -> Optional[Dict[str, any]]:
     """
     Write a new file from a HTTP multipart request.
 
+    :param req: aiohttp request object
+    :param file_path: Either a path to a folder, or a complete path that includes a filename.
+    :param compressed: Whether or not to enforce gzip compression, defaults to False
     :return: size of the new file in bytes
     """
     reader = await req.multipart()
     file = await reader.next()
+    new_file = dict()
 
     size = 0
 
+    if not path.suffix:
+        path = path / file.filename
+
     try:
-        await req.app["run_in_thread"](os.makedirs, file_path.parent)
+        await req.app["run_in_thread"](os.makedirs, path.parent)
     except FileExistsError:
         pass
 
-    async with aiofiles.open(file_path, "wb") as handle:
+    async with aiofiles.open(path, "wb") as handle:
         while True:
             chunk = await file.read_chunk(CHUNK_SIZE)
             if not chunk:
                 break
+
+            if size == 0 and compressed:
+                if not is_gzip_compressed(chunk):
+                    return None
+
             size += len(chunk)
             await handle.write(chunk)
 
-    return size
+    new_file.update(
+        {
+            "name_on_disk": path.name,
+            "size": size,
+            "uploaded_at": virtool.utils.timestamp(),
+        }
+    )
```

Comment:
It is best to limit the scope of `naive_writer` to writing the file and returning values that are directly related to the writing of the file (`size`). Limit functions to one responsibility wherever possible.

You pass in the target `path` to `naive_writer` so that calling scope already knows the `path` and can calculate `name_on_disk` if it sees fit. The timestamp can also be easily calculated (or not) in the calling scope.

It is not a good idea to return a `dict` like you did above. They are a really fuzzy way of passing around data. There is no guarantee that the expected keys have been defined in the code that created the `dict` and it is impossible to type individual values. One solution is to gather your values in an object or dataclass that makes logical sense.

If you have a small number of values, return a tuple:
```
return name, size
```

This is much easier to type and an exception will be raised if a mismatch in the number of items in the tuple does not match. https://stackoverflow.com/questions/1479776/too-many-values-to-unpack-exception

Reply:
> You pass in the target `path` to `naive_writer` so that calling scope already knows the `path` and can calculate `name_on_disk` if it sees fit. The timestamp can also be easily calculated (or not) in the calling scope.

Right now it is set up to work with reads uploads (where `naive_writer` is only given the path to the folder) so the name of the file is determined by reading the multipart, then a row is created using that. 

If I were to use a required `name` query I can get away with not returning the name, but then `name` and `name_on_disk` would be the exact same (since `name_on_disk`  isn't something like `{id}-reads_1.fq.gz`).